### PR TITLE
TestCase for BeatSyncedContainer

### PR DESCRIPTION
**Pull request by EVAST9919**

### CODE REVIEW

**Comment by peppy** [comment]
what is a `beatLifeTime`? do you mean `beatLength`?

**Comment by EVAST9919** [comment]
Yes, I mean that, but it's a bit different than `timingPoint.BeatLength`. So should I name it `beatLength` anyway?

File: osu.Game/Graphics/Containers/BeatSyncedContainer.cs [hunk 1]
```diff
@@ -42,25 +45,49 @@ protected override void Update()
...
+            if (beatIndex < 0)
+                return;
+
             if (timingPoint == lastTimingPoint && beatIndex == lastBeat)
                 return;
 
             double offsetFromBeat = (timingPoint.Time - currentTrackTime) % timingPoint.BeatLength;
 
+            // Beat can be trimmed by the next timing point, so in some cases we can't use timingPoint.BeatLength
+            double beatLength = calculateBeatLength(timingPoint, getNextTimingPoint(timingPoint), currentTrackTime);
```
Comment (peppy) [inline osu.Game/Graphics/Containers/BeatSyncedContainer.cs]: as i mentioned in discord, i don't believe we *want* to clip the beatlength in cases like this. it creates uneven effects.

File: osu.Desktop.VisualTests/Tests/TestCaseBeatSyncedContainer.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                        AutoSizeAxes = Axes.Both,
+                        Margin = new MarginPadding { Bottom = flash_layer_heigth, },
+                        Children = new Drawable[]
+                        {
+                            timingPointCount = new InfoString(@"Timing points amount: "),
+                            currentTimingPoint = new InfoString(@"Current timing point: "),
+                            beatCount = new InfoString(@"Beats amount (in the current timing point): "),
+                            currentBeat = new InfoString(@"Current beat: "),
+                            beatsPerMinute = new InfoString(@"BPM: "),
+                            beatLength = new InfoString(@"Beat length: "),
```
Comment (peppy) [inline osu.Desktop.VisualTests/Tests/TestCaseBeatSyncedContainer.cs]: call this "adjusted beat length" or something (and show non-adjusted).

also remove spaces from the end of these and add them in the class you made for it

File: osu.Desktop.VisualTests/Tests/TestCaseBeatSyncedContainer.cs [hunk 3]
```diff
@@ -77,12 +77,12 @@ public BeatContainer()
...
-                            beatCount = new InfoString(@"Beats amount (in the current timing point): "),
-                            currentBeat = new InfoString(@"Current beat: "),
-                            beatsPerMinute = new InfoString(@"BPM: "),
-                            beatLength = new InfoString(@"Beat length: "),
+                            timingPointCount = new InfoString(@"Timing points amount"),
+                            currentTimingPoint = new InfoString(@"Current timing point"),
+                            beatCount = new InfoString(@"Beats amount (in the current timing point)"),
+                            currentBeat = new InfoString(@"Current beat"),
+                            beatsPerMinute = new InfoString(@"BPM"),
+                            adjustedBeatLength = new InfoString(@"Beat length"),
```
Comment (peppy) [inline osu.Desktop.VisualTests/Tests/TestCaseBeatSyncedContainer.cs]: but the display string is still beat length, which is wrong

**Review by peppy — APPROVED** [review]

**Pull request merged**